### PR TITLE
Haskell/explicit fees

### LINKED ISSUE #100: Explicit Fees (executable model)
The executable model currently uses implicit fees. Make them explicit and validate a minimum fee.

### PR DESCRIPTION
**Pull request by JaredCorduan**
This add explicit fee to the executable model.  It also enforces a minimum fee.

closes #100

### CODE REVIEW

File: hs/src/PtrlConsts.hs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module PtrlConsts
```
Comment (JaredCorduan) [inline hs/src/PtrlConsts.hs]: I just noticed the misspelling.  I'll fix this along with anything else that comes up :)

File: hs/src/LedgerState.hs [hunk 2]
```diff
@@ -58,8 +59,10 @@ data LedgerValidation = LedgerValidation [ValidationError] LedgerState
 data ValidationError =
   -- | The transaction inputs are not valid.
     BadInputs
-  -- | The transaction results in an increased total balance of the ledger.
-  | IncreasedTotalBalance
+  -- | The transaction fee is too small
+  | FeeTooSmall String
```
Comment (mgudemann) [inline hs/src/LedgerState.hs]: I would prefer an integer argument here, I understand that just the difference between minimum and actual might not be enought, but why not `FeeTooSmall $MINIMUM $ACTUAL` with two `Natural` parameters?

File: hs/src/LedgerState.hs [hunk 3]
```diff
@@ -155,13 +161,29 @@ validInputs (TxWits tx _) l =
     else Invalid [BadInputs]
   where unspentInputs (UTxO utxo) = Map.keysSet utxo
 
+-- |Minimum fee calculation
+minfee :: PtrlConsts -> TxWits -> Coin
+minfee pc txwit = minfeeA pc * x + minfeeB pc
+  where x = Coin $ toEnum $ length (show txwit)
```
Comment (mgudemann) [inline hs/src/LedgerState.hs]: I think we should add this reasoning and actual formula for this in the documentation of the function. This will definitely influence test cases, generators etc.

File: hs/src/PtrlConsts.hs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    ) where
+
+import Coin (Coin(..))
+
+data PtrlConsts =
+  PtrlConsts
+  { -- |The linear factor for the minimum fee calculation
+    minfeeA :: Coin
+    -- |The constant factor for the minimum fee calculation
+  , minfeeB :: Coin
```
Comment (mgudemann) [inline hs/src/PtrlConsts.hs]: Why are these of type `Coin` ?

File: hs/test/Tests.hs [hunk 5]
```diff
@@ -202,21 +214,23 @@ testSpendNonexistentInput =
...
     txbody = Tx
               (Set.fromList [TxIn genesisId 42])
-              [ TxOut aliceAddr (Coin 0) ]
+              [ TxOut aliceAddr (Coin 10000) ]
               Set.empty
+              (Coin 1500)
     aliceWit = makeWitness alicePay txbody
     tx = TxWits txbody (Set.fromList [aliceWit])
-  in ledgerState [tx] @?= Left [BadInputs, InsufficientWitnesses]
+  in ledgerState [tx] @?= Left [BadInputs, ValueNotConserved, InsufficientWitnesses]
```
Comment (mgudemann) [inline hs/test/Tests.hs]: This changes the semantics of the test case. Shouldn't this be at least in a separate commit?

File: hs/src/LedgerState.hs (hunk 3 above)
Comment (JaredCorduan) [reply hs/src/LedgerState.hs]: in the spec I left it abstract, since I didn't think it mattered to much what the calculation was.  here I based it loosely on how it is done today (where `length (show txwit)` is just a simple bytes-like calculation).  But I'm happy making it explicit in the spec, or at least saying it's this ax+b calculation (and not worrying about an actual encoding here).

File: hs/src/PtrlConsts.hs (hunk 4 above)
Comment (JaredCorduan) [reply hs/src/PtrlConsts.hs]: is it odd? `minfeeA` would be the number of lovelace per byte that you must pay as a fee, and `minfeeB` is an additional constant number of lovelace as a constant factor. would it be more natural to convert to `Coin` only at the end?

File: hs/test/Tests.hs (hunk 5 above)
Comment (JaredCorduan) [reply hs/test/Tests.hs]: changing the preservation of value from and inequality to an equality changes the semantics. it causes some correlation between the errors that is different than how it was before. I'm not sure how to make the equality change without also making a test change (or have tests broken between commits).  Or am I missing something?

File: hs/src/PtrlConsts.hs (hunk 1 above)
Comment (JaredCorduan) [reply hs/src/PtrlConsts.hs]: fixed

File: hs/src/LedgerState.hs (hunk 2 above)
Comment (JaredCorduan) [reply hs/src/LedgerState.hs]: I made it two `Coin`s.

File: hs/test/Tests.hs (hunk 5 above)
Comment (mgudemann) [reply hs/test/Tests.hs]: you are perfeclty right. For me, it wouldn't be bad to have a test failing in one commit if it's fixed within the next commit. But that might be problematic for bisection maybe? So leave it like it is if you want to.

File: hs/src/PtrlConsts.hs (hunk 4 above)
Comment (mgudemann) [reply hs/src/PtrlConsts.hs]: hm, I had this in mind as general parameters, e.g., parameters for a linear function like `y=a*x + b` where type of `Coin` for `a` would not make much sense, right?

Comment (JaredCorduan) [reply hs/src/PtrlConsts.hs]: I'm happy either way. In general, I try not to pull out the underlying nat inside the coin, so thinking of the coin as a value. Then the type of `a` and `b` would have to be `Coin`. And it makes sense to me that `a` is the number of coins per byte. But maybe that's just me, I don't know. :)

File: hs/src/LedgerState.hs (hunk 3 above)
Comment (JaredCorduan) [reply hs/src/LedgerState.hs]: I made an issue: #102
